File: ai_lab/gateway.py
"""Model routing, admission and load coordination for agent requests.

Scheduling, memory accounting, request state and telemetry live in their
respective modules. See docs/gateway-behavior.md for the operating contract.
"""

from __future__ import annotations

import time

from . import budget
from .operations import Operations
from .eviction import EvictionPlanner
from .gateway_stats import GatewayCounters, GatewayStats
from .gateway_state import Lease, Shape
from .gateway_errors import CardBusy, CouldNotLoad, NotConfigured, ShapeNotServed
from .gateway_resources import GatewayResources
from .gateway_control import GatewayControl
from .scheduler import Abandoned, Scheduler, WillNotFit
from .types import Task


# How quiet the card has to be before a new model is loaded, and how long to
# wait for that. The threshold is not zero because a driver keeps a small
# allocation of its own; on this machine an idle card reads about 2 MiB.
QUIET_MB = 512.0
QUIET_TIMEOUT_S = 60.0
QUIET_POLL_S = 0.5

# How long to wait for an engine, and how many requests to hold. All three are
# defaults: the real values live in the configuration, because the right
# numbers differ between a card that reads 8,400 tokens of prompt in under a
# second and a Mac running a 70 GB model at 17 tokens a second.
FIRST_BYTE_S = 120.0
BETWEEN_BYTES_S = 30.0
MAX_WAITING = 150


class Gateway:
    """Routes by model name and puts that model on the card if it is not there.

    The queueing itself is `scheduler.Scheduler`, which knows nothing about
    models: it is handed a way to put a shape on the card and a way to ask how
    many requests that shape serves at once. Everything about names, engines
    and settings is here; everything about who goes next is there.
    """

    HISTORY = 50

    def __init__(self, operations: Operations,
                 quiet_mb: float = QUIET_MB,
                 quiet_timeout_s: float = QUIET_TIMEOUT_S,
                 poll_s: float = QUIET_POLL_S,
                 first_byte_s: float = FIRST_BYTE_S,
                 between_bytes_s: float = BETWEEN_BYTES_S,
                 max_waiting: int = MAX_WAITING,
                 task_timeouts: dict | None = None,
                 max_upload_bytes: int = 0,
                 max_upload_pixels: int = 0,
                 max_upload_dimension: int = 0) -> None:
        self.operations = operations
        self.quiet_mb = quiet_mb
        self.quiet_timeout_s = quiet_timeout_s
        self.poll_s = poll_s
        self.first_byte_s = first_byte_s
        self.between_bytes_s = between_bytes_s
        # Per-task overrides of the two waits above, keyed by `Task.value`.
        # A slow image workflow and a fast text completion share one gateway
        # but must not share one timeout: the default alone is sized for text.
        # Missing keys, or a task with no override, fall back to the defaults.
        self.task_timeouts = dict(task_timeouts or {})
        self.max_upload_bytes = max_upload_bytes
        self.max_upload_pixels = max_upload_pixels
        self.max_upload_dimension = max_upload_dimension
        # What the last card reading said, per pool. Refreshed around every
        # load, never from inside the scheduler's lock.
        self._budget_pools: dict = {}
        self.resources = GatewayResources(operations, quiet_mb,
                                          quiet_timeout_s, poll_s)
        self.eviction = EvictionPlanner()
        self.scheduler = Scheduler(self._put_on_card, self._places,
                                   self._make_room,
                                   max_waiting=max_waiting)
        # Whether anything outside may have changed what is on the card. Set at
        # startup and whenever a button on the page loads or unloads something.
        # The next switch sweeps: one expensive read after an outside change
        # rather than one on every request.
        self._resweep = True
        self.control = GatewayControl(self.scheduler)
        self.counters = GatewayCounters()

    # -- what a client can ask for -----------------------------------------

    def catalogue(self) -> list[dict]:
        """Every configured entry, loaded or not, with the names it answers to.

        Entries that are not loaded are listed too. A client is meant to be able
        to ask for one of them — that is the whole point.
        """
        return [{
            "id": instance["id"],
            "model_id": instance["model_id"],
            "engine": instance["engine"],
            "port": instance["port"],
            "loaded": bool(instance["running"]),
            "ready": bool(instance["ready"]),

            # Which shapes of request this one answers. A client that speaks
            # only one of them can tell from the listing which models are open
            # to it, instead of finding out by being refused.
            "shapes": self._shapes(instance),
        } for instance in self.operations.instances()]

    def _shapes(self, instance: dict) -> list[str]:
        try:
            task = Task(instance.get("task", Task.TEXT_GENERATION.value))
            return list(self.operations.engines.get(instance["engine"]).api_paths(task))
        except KeyError:
            return []

    @staticmethod
    def _engine_name(instance: dict) -> str:
        """What this entry's engine calls its own model.

        The last segment of the model path, which is the name the engine is
        started with and the only one it will answer to.
        """
        model_id = instance.get("model_id") or ""
        return model_id.rsplit("/", 1)[-1]

    def resolve(self, wanted: str, instances: list[dict] | None = None) -> dict:
        """The entry with this id, or NotConfigured naming what is known.

        The id and nothing else. It used to answer to four names — the id, the
        label a person gave it, the model's path, and the file at the end of
        that path — and the first match won.

        That is a collision waiting to be found. Two entries pointing at one
        model with different settings is not a strange thing to want; it is
        exactly what the settings in a request are for. Both would have
        answered to the file's name, one of them would have won silently, and
        the request would have been served by the wrong one.

        So there is one name. It is stable, it survives renaming the label, it
        has no spaces in it, and it fits the single `model` field the request
        shapes give us. The label is for reading; this is for sending.
        """
        key = (wanted or "").strip().lower()
        if instances is None:
            instances = self.operations.configured()
        for instance in instances:
            if instance["id"].strip().lower() == key:
                return instance
        known = sorted(instance["id"] for instance in instances)
        raise NotConfigured(
            f"No configured model answers to {wanted!r}. Known: {', '.join(known)}")

    # -- taking a place on the card -----------------------------------------

    def acquire(self, wanted: str, shape: str | None = None,
                settings: dict | None = None,
                still_wanted=None) -> Lease:
        """Take a place on the card, with the named model on it.

        Returns once the model is answering and this request may go through.
        It may return at once — the model is loaded and there is room — or after
        waiting for a place, or after waiting for a switch and the load.

        `shape` is the kind of request about to be forwarded. Checked before
        anything is queued: an entry whose engine does not answer that shape is
        refused straight away rather than after a forty-second load.

        `settings` asks for the model started with something other than what the
        entry is configured with. Checked before queueing too, so a setting the
        engine does not have comes back at once. It is part of what is being
        asked for: two requests wanting different context sizes cannot share a
        card, whatever the model.

        `still_wanted()` says whether the client is still there. Asked when this
        request reaches the head of the queue and before anything is unloaded.
        A client that gave up must not cost a swap — reproduced on the machine
        before this existed: one hung up at once and the manager took a working
        model off the card to load twenty-one gigabytes for nobody.
        """
        started = time.perf_counter()
        # The configuration, not the supervisor. Which entry answers to a name,
        # which engine runs it, what settings it has — all of that is the file,
        # at 0.05 ms. Asking what every instance is *doing* costs 73 ms on the
        # container, and this path needs that only when something outside may
        # have changed the card.
        entries = self.operations.configured()
        instance = self.resolve(wanted, entries)
        if shape is not None:
            self._refuse_wrong_shape(instance, shape, entries)
        instance_id, port = instance["id"], instance["port"]
        # Always the full settings, never only what was asked for. A request
        # naming no settings and one naming exactly the configured ones are
        # asking for the same thing, and comparing partial dictionaries would
        # make them different — a reload for nothing, on every other request.
        asked_for = self.operations.effective_params(instance_id, settings or {})
        self._adopt_what_is_there()
        requested = Shape.of(instance_id, asked_for)
        # Some engines load weights only on their first inference. The card
        # reading taken when the server started is then too optimistic for the
        # next model. Refresh only for a switch; warm requests stay cheap.
        if not any(item["shape"] == requested
                   for item in self.scheduler.state()["loaded"]):
            self._read_the_card()

        try:
            self.scheduler.enter(requested,
                                 still_wanted=still_wanted)
        except WillNotFit as error:
            # The scheduler knows the ordering and nothing about memory, so the
            # numbers are attached here. A client that can read them can correct
            # itself — a smaller context, a smaller share of the card — without
            # anybody parsing an English sentence.
            error.detail = {"ai_lab": self._why_it_does_not_fit(
                instance_id, asked_for)}
            self.counters.last_error = str(error)
            raise
        except Exception as error:
            # A client that gave up is not a fault of this machine, so it does
            # not become the error the page shows.
            if not isinstance(error, Abandoned):
                self.counters.last_error = str(error)
            raise
        try:
            self.counters.requests += 1
            self.counters.waited_s += time.perf_counter() - started
            self.counters.arrivals.append((time.time(), instance_id))
            return Lease(self, instance_id, port, self._engine_name(instance),
                         started=time.perf_counter())
        except BaseException:
            # The place has been taken. Anything that goes wrong between there
            # and handing it to the caller has to give it back, or the card
            # loses a place with nobody using it.
            self.scheduler.leave()
            raise

    def finished(self, held_s: float = 0.0) -> None:
        """Give a place back, and let in whoever can go next."""
        self.counters.served_s += held_s
        self.scheduler.leave()

    def first_token(self, seconds: float, instance_id: str = "") -> None:
        """How long that request waited for its first token.

        Reported only for requests that asked for streaming — see `GatewayCounters`.
        Kept against the model that answered, because that is the only level at
        which the figure means anything.
        """
        counters = self.counters
        counters.first_token_s[instance_id] = (
            counters.first_token_s.get(instance_id, 0.0) + seconds)
        counters.first_tokens[instance_id] = (
            counters.first_tokens.get(instance_id, 0) + 1)

    def _adopt_what_is_there(self) -> None:
        """Tell the scheduler what is already on the card, once.

        Only after something outside may have changed it — a manager that has
        just started, or a button on the page. Otherwise the scheduler is the
        authority and reading again would only be a chance to disagree with
        itself.

        Everything that is up and answering, however many. This used to insist
        on exactly one and give up otherwise — two engines up was not a card to
        adopt but a card to clear — and with a memory budget that is simply
        the ordinary state. The page reported "nothing loaded" while two models
        were serving.

        A model that is up but not yet answering is left for later, and the
        flag stays set so the question is asked again. Adopting it would send
        requests to a port with nothing behind it; giving up on it for good
        would leave it invisible until something else changed.
        """
        if not self._resweep:
            return
        # The expensive question, asked once after an outside change rather
        # than on every request.
        instances = self.operations.instances()
        running = [item for item in instances if item["running"]]
        if any(not item["ready"] for item in running):
            return                      # still coming up: ask again next time
        self._resweep = False
        self.scheduler.adopt(*[
            Shape.of(item["id"], {**item.get("params", {}),
                                  **item.get("active_params", {})})
            for item in running])

    # -- what the scheduler asks of us --------------------------------------

    def _places(self, shape: "Shape | None") -> int:
        """How many requests this shape serves at once. The engine's own number.

        One when it cannot be worked out — a shape naming an entry that has
        been deleted, say. Guessing high there would let requests through to a
        model that cannot take them.
        """
        if shape is None:
            return 1
        try:
            instance = self.operations.instance(shape.instance_id)
            engine = self.operations.engines.get(instance["engine"])
            return max(1, engine.concurrency({**instance["params"],
                                              **shape.as_dict()}))
        except Exception:
            return 1

    # -- deciding who has to go --------------------------------------------

    def _make_room(self, shape: "Shape", loaded: list[dict]) -> "list | None":
        """Choose victims from cached memory figures under the scheduler lock."""
        return self.eviction.choose(
            shape, loaded, needs_mb=self._needs_mb,
            free_mb=self._free_mb(), capacity_mb=self._capacity_mb())

    def _why_it_does_not_fit(self, instance_id: str, asked_for: dict) -> dict:
        """The numbers behind a refusal, for a client to act on.

        Everything here is a fact about this machine right now. Nothing is
        remembered between requests: what a model took last time is knowledge
        that belongs to whatever is making the requests, not to this manager.
        """
        shape = Shape.of(instance_id, asked_for)
        pool = (self._budget_pools.get(budget.CARD)
                or self._budget_pools.get(budget.MACHINE) or {})
        return {
            "model": instance_id,
            "engine": self._engine_name_of(shape),
            "needed_mb": round(self._needs_mb(shape)),
            "capacity_mb": round(self._capacity_mb()),
            "available_mb": round(self._free_mb()),
            "pool": pool.get("name", ""),
            "loaded": [item.instance_id for item in self._scheduled_shapes()],
            "asked": dict(asked_for),
        }

    def _needs_mb(self, shape: "Shape") -> float:
        return self.resources.needs_mb(shape, self._card_total_mb())

    def _card_total_mb(self) -> float:
        return self.resources.card_total_mb()

    def _capacity_mb(self) -> float:
        return self.resources.capacity_mb(self._budget_pools)

    def _free_mb(self) -> float:
        return self.resources.free_mb(self._budget_pools)

    def _read_the_card(self) -> None:
        # Never called while holding the scheduler lock. Keep the last good
        # reading if the host probe fails during a model transition.
        try:
            self._budget_pools = self.resources.read_pools()
        except Exception:
            pass

    def _put_on_card(self, shape: "Shape", victims: tuple = ()) -> None:
        """Take those off, wait for the card to go quiet, then load this.

        `victims` is what the scheduler decided has to go. Empty means this
        fits beside what is already there and nothing is disturbed.

        A manager restart or a machine boot can leave engines up that nothing
        here knows about. Those are cleared too, on the first switch after such
        a change — see `_resweep` — because loading into whatever is left is
        how a model that is known to fit fails to fit.
        """
        started = time.perf_counter()
        unloaded = self._clear(victims)
        self._resweep = False
        self._wait_until_quiet(unloaded)
        self._read_the_card()
        self._refuse_if_still_full(shape)

        operation = self.operations.load(shape.instance_id, shape.as_dict() or None)
        if not operation.ok:
            raise CouldNotLoad(
                operation.error or f"{shape.instance_id} would not start")

        self._read_the_card()
        took = time.perf_counter() - started
        self.counters.switches += 1
        if unloaded:
            self.counters.evictions += len(unloaded)
        self.counters.switch_s += took
        self.counters.history.append({
            "at": time.time(), "loaded": shape.instance_id, "unloaded": unloaded,
            "took_s": round(took, 1), "load_ms": operation.total_ms,
        })
        del self.counters.history[:-self.HISTORY]

    def _clear(self, victims: tuple = ()) -> list[str]:
        """Unload what was chosen, and anything nothing here knows about.

        The second half is the point of sweeping rather than unloading by name:
        a manager restart can leave an engine running that the scheduler never
        adopted, and its memory is real whether or not this knows about it.
        """
        wanted = {shape.instance_id for shape in victims}
        known = {item.instance_id for item in self._scheduled_shapes()}
        stopped = []
        for instance in self.operations.instances():
            if not instance["running"]:
                continue
            # The chosen victims, and anything running that the scheduler
            # does not believe in. A stray is not in the budget arithmetic, so
            # its memory is unaccounted — and loading into what is left is how
            # a model that is known to fit fails to.
            if instance["id"] in wanted or instance["id"] not in known:
                self.operations.unload(instance["id"])
                stopped.append(instance["id"])
        return stopped

    def _scheduled_shapes(self) -> list:
        """What the scheduler believes is loaded."""
        return [item["shape"] for item in self.scheduler.state()["loaded"]]

    # -- how long to wait, per kind of request --------------------------

    def timeouts_for(self, task: Task) -> tuple[float, float]:
        """(first_byte_s, between_bytes_s) for this task.

        An override for `task.value` in `task_timeouts` wins; anything not
        named there gets the machine's default pair. Text generation is not
        given its own entry — it *is* the default, since every number above
        was sized against it — but a caller may still add one.
        """
        override = self.task_timeouts.get(task.value, {})
        return (float(override.get("first_byte_s", self.first_byte_s)),
                float(override.get("between_bytes_s", self.between_bytes_s)))

    # -- what the buttons on the page have to tell us -----------------------

    def apply_settings(self, settings: dict) -> None:
        """Take new limits without a restart.

        The queue length is the scheduler's; the two waits are used by the web
        layer when it forwards, and it reads them from here each time, so a
        change reaches the next request rather than the next restart.
        """
        if "first_byte_s" in settings:
            self.first_byte_s = float(settings["first_byte_s"])
        if "between_bytes_s" in settings:
            self.between_bytes_s = float(settings["between_bytes_s"])
        if "max_waiting" in settings:
            self.scheduler.max_waiting = int(settings["max_waiting"])
        if "task_timeouts" in settings:
            self.task_timeouts = dict(settings["task_timeouts"])

    def card_changed(self) -> None:
        """Something outside loaded or unloaded a model.

        The scheduler believes what it put there. A button on the page can put
        something else, and the next request would then be admitted onto a card
        that no longer holds what it thinks.
        """
        self.scheduler.forget_current()
        self._resweep = True

    def reset(self, reason: str) -> int:
        """Throw everything away. What a forced stop means.

        Everything in flight is already dying — whoever forced it killed the
        engine. Everyone waiting is turned away rather than left queueing for a
        state that no longer exists. Returns how many were turned away.
        """
        self._resweep = True
        return self.scheduler.reset(reason)

    # -- which shapes an entry answers --------------------------------------

    def _answers(self, instance: dict, shape: str) -> bool:
        """Whether this entry's engine answers this kind of request."""
        try:
            engine = self.operations.engines.get(instance["engine"])
        except KeyError:
            return False
        task = Task(instance.get("task", Task.TEXT_GENERATION.value))
        return shape in engine.api_paths(task)

    def _refuse_wrong_shape(self, instance: dict, shape: str,
                            instances: list[dict]) -> None:
        """Say no, and say which entries would have worked.

        A client that sent the wrong shape does not know which of its models
        are on which engine, and has no way to find out from a refusal that
        only says no. Naming them turns a dead end into one edit.
        """
        if self._answers(instance, shape):
            return
        able = sorted(other["id"] for other in instances
                      if self._answers(other, shape))
        raise ShapeNotServed(
            f"{instance['id']} runs on {instance['engine']}, which does not "
            f"answer {shape}. "
            + (f"Configured models that do: {', '.join(able)}." if able
               else "No configured model answers it."))

    # -- what the buttons on the page have to respect -----------------------

    def busy(self) -> dict | None:
        return self.control.busy()

    def guard(self, action: str, instance_id: str) -> None:
        self.control.guard(action, instance_id)

    def _wait_until_quiet(self, stopped: list) -> None:
        self.resources._wait_until_quiet(stopped)

    def _wait_for_release(self, deadline: float, keeping: bool) -> None:
        self.resources._wait_for_release(deadline, keeping)

    def _refuse_if_still_full(self, shape: "Shape") -> None:
        """Say so before starting an engine that cannot fit.

        Asked only when the arithmetic is available. Where it is not, the
        machine was emptied instead and `_wait_for_release` has already
        insisted the card reached nothing.
        """
        needed = self._needs_mb(shape)
        free = self._free_mb()
        if not needed or not free or needed <= free:
            return
        raise CouldNotLoad(
            f"{shape.instance_id} needs about {needed / 1024:.1f} GB and "
            f"{free / 1024:.1f} GB is free after unloading what could be "
            f"unloaded. Something outside AI-Lab is holding the card, or an "
            f"engine did not release its memory.")

    def _loaded(self) -> str | None:
        """What is on the card, read rather than remembered.

        A remembered value goes stale the moment something outside the gateway
        unloads a model — and that happens on purpose: a person can force past
        a busy card from the page, and the gateway is deliberately not told.
        Reading it costs one call the caller is making anyway, and it cannot be
        wrong.

        One name, and the first one found. This is used where a single answer
        is wanted — the page asks the scheduler for the whole set. Anything
        running that the scheduler does not know about is swept off by the next
        switch, because its memory is real and unaccounted for.
        """
        for instance in self.operations.instances():
            if instance["running"]:
                return instance["id"]
        return None

    # -- what the interface shows ------------------------------------------

    def stats(self) -> dict:
        return GatewayStats(self).snapshot()

    def _engine_name_of(self, shape) -> str:
        if shape is None:
            return ""
        try:
            entry = self.operations.instance(shape.instance_id)
            return self.operations.engines.get(entry["engine"]).display_name
        except Exception:
            return ""

    def _shapes_offered(self) -> list[dict]:
        return GatewayStats(self)._shapes_offered()

File: ai_lab/web/js/views/gateway.js
// The one address an agent talks to, and what it has been doing.
//
// The numbers here answer one question: is the workflow running, or is it
// spending its time changing models? A switch is an unload, a wait for the
// card to go quiet, and a load — tens of seconds. When switches are close to
// requests, almost every step changes model, and reordering the workflow so
// that steps using the same model sit together is worth more than any setting
// on this page.
//
// This is the only view that redraws on a timer. Everywhere else a change
// comes from something the server did and arrives as an event; here the
// traffic comes from outside the manager entirely and produces no events at
// all, so a page that never redrew would sit at whatever the numbers were when
// the tab was opened.

import { api } from '../api.js';
import { element, seconds } from '../format.js';
import { showNotice } from '../confirm.js';

// Slower than it looks like it should be, on purpose. Reading these numbers
// means asking systemd about every configured instance and probing each one
// that is up: measured on the container with eleven instances, one call costs
// about 125 ms and spawns 28 processes. And the thing being watched moves far
// more slowly than that — a model switch takes between twelve and forty-five
// seconds. A two-second poll was finer than its own subject and cost fourteen
// processes a second for the privilege.
const EVERY_MS = 5000;
let timer = null;

// A heading above its panel, the same shape the Settings page uses.
function section(title, children) {
  return element('div', { class: 'section' }, [
    element('h3', { text: title }),
    element('div', { class: 'card' }, children),
  ]);
}

function line(label, value, title) {
  return element('div', { class: 'row', ...(title ? { title } : {}) }, [
    element('span', { class: 'muted', text: label }),
    element('span', { text: value }),
  ]);
}

// The same, for something that is not a measurement. An address and a key are
// things to read and copy, not figures to compare, so they do not take the
// emphasis the numbers get.
function plain(label, value, title) {
  return element('div', { class: 'row', ...(title ? { title } : {}) }, [
    element('span', { class: 'muted', text: label }),
    element('span', { class: 'muted', text: value }),
  ]);
}

// Where to point an agent. Built from the page being looked at rather than
// from the server's idea of itself: the manager may be reached by name, by
// address, or through a tunnel, and the answer has to be the address that
// actually worked to get here.
function address(stats) {
  const base = `${window.location.protocol}//${window.location.host}`;
  const all = stats.shapes.reduce((most, row) =>
    Math.max(most, row.models.length), 0);

  // Two shapes of request are in circulation, and listing only the address
  // would be half the answer — the wrong half for anybody whose tool speaks
  // the other one. Grouped by which models answer each, because that is the
  // part that differs: every engine answers the usual shape, and only some
  // answer Anthropic's.
  const shapes = stats.shapes.map((row) => {
    const everyone = row.models.length === all;
    // The engine, not the models. A list of names grows every time an entry is
    // added and is stale by the next one; "vLLM models" says it once and stays
    // true.
    const who = everyone ? 'every model'
      : `${row.engines.join(' and ')} models`;
    return element('div', { class: 'row tight', title: everyone
      ? 'Every configured model answers this'
      : `${row.models.length} of ${all}: ${row.models.join(', ')}` }, [
      element('span', { class: 'muted', text: row.path }),
      element('span', { class: 'muted', text: who }),
    ]);
  });

  return section('Server address', [
    plain('Base URL', `${base}/v1`),
    ...shapes,
  ]);
}

// What the accelerator says about itself. Memory is the binding constraint on
// the binding constraint on how many models fit; the other two arrive in the same
// reading, so they cost nothing extra.
// What the machine has room for, and how warm it is.
//
// The pools come from the same reading that decides whether a model may be
// loaded, so this page cannot disagree with the thing that says no. On a
// dedicated card there are two — the card, used whole, and the machine, which
// keeps a reserve. On Apple silicon there is one, because the chip and
// everything else draw on the same memory.
const POOL_NAMES = { card: 'Card mem', machine: 'System mem' };

// A figure, or a dash. Every number here comes from a manager that may be
// older than this file — they are not replaced in the same instant — and
// `String(undefined)` reaches the screen as the word "undefined". A dash says
// "no answer", which is true, and is what the rest of this interface uses.
function figure(value, suffix = '') {
  const found = Number(value);
  return Number.isFinite(found) ? `${found}${suffix}` : '—';
}

function memoryLines(memory, card) {
  const lines = [];
  if (card && card.temperature_c !== null && card.temperature_c !== undefined) {
    lines.push(line('GPU temp', figure(card.temperature_c, ' °C')));
  }
  (memory && memory.pools ? memory.pools : []).forEach((pool) => {
    const share = pool.total_mb
      ? Math.round((100 * pool.used_mb) / pool.total_mb) : 0;
    lines.push(line(POOL_NAMES[pool.name] || pool.name,
                    `${figure(pool.used_mb)} / ${figure(pool.total_mb)} MB`,
                    pool.kind === 'unified'
                      ? `${share}% in use, by models and by everything else. `
                        + 'Apple silicon shares one pool between the chip and '
                        + 'the rest of the machine.'
                      : pool.name === 'card'
                        ? `${share}% of the card. All of it is for models.`
                        : `${share}% in use. ${pool.reserve_mb} MB of the rest `
                          + 'is held back for the machine itself.'));
  });
  if (memory && memory.pools && memory.pools.length) {
    lines.push(line('Available mem', figure(memory.available_mb, ' MB'),
                    'What is free, less what is held back for the machine. A '
                    + 'model has to fit in one pool, so this total is the '
                    + 'ceiling rather than the test.'));
  }
  return lines;
}

function activity(stats) {
  const loaded = stats.loaded || [];
  const status = stats.switching ? 'loading a model'
    : stats.in_flight && stats.waiting ? 'working, with a queue'
    : stats.in_flight ? 'working'
    : stats.waiting ? 'waiting for room'
    : 'idle';
  const places = loaded.reduce((total, item) => total + (Number(item.places) || 0), 0);
  return section('Activity', [
    line('Status', status),
    line('Loaded', figure(loaded.length),
         'How many models are on the machine. Which ones, and what each is '
         + 'doing, is below.'),
    line('Processing', `${figure(stats.in_flight)} from ${figure(places)}`,
         'Requests running right now, against the places every loaded model '
         + 'offers between them. Places are the engine\u2019s own number and '
         + 'differ per model.'),
    line('Queue size', figure(stats.waiting),
         'Requests waiting for a model that is not loaded.'),
    line('Requests per minute', figure(stats.requests_per_minute),
         'In the last sixty seconds.'),
    line('Switches', figure(stats.switches)),
    line('Time spent switching', figure(stats.switching_share, '%'),
         'Of the time this was working — answering or loading — how much went '
         + 'on loading.'),
    ...memoryLines(stats.memory, stats.card),
    stats.last_error
      ? element('p', { class: 'error', text: `Last error: ${stats.last_error}` })
      : null,
  ].filter(Boolean));
}

// The three numbers that decide how patient the front door is, and how much it
// will hold. Editable here rather than on the Settings page: they are about
// the gateway, and they sit beside the figures you would read before deciding
// to change them.
function limits(stats, redraw) {
  const fields = [
    ['first_byte_s', 'Max TTFT', stats.first_byte_s, 's',
     'How long to wait for an engine to start answering. It covers reading the '
     + 'prompt — and the whole answer for a request that did not ask for '
     + 'streaming, since such an engine sends nothing at all until it has '
     + 'finished writing. A large prompt on a slow machine is the case to size '
     + 'this for.'],
    ['between_bytes_s', 'Max idle time', stats.between_bytes_s, 's',
     'How long a silence in the middle of an answer means the engine has '
     + 'stopped rather than slowed. At the slowest generation measured here, '
     + '17 tokens a second, the gap between them is 59 milliseconds. It only '
     + 'applies to a request that asked for streaming: without it the answer '
     + 'arrives in one piece and there are no gaps to measure.'],
    ['max_waiting', 'Max queue size', stats.max_waiting, '',
     'How many may wait for a model at once. Beyond it a request is refused '
     + 'rather than queued, because each one waiting occupies a thread. A '
     + 'workflow that hits this is asking for more at once than one card can '
     + 'answer.'],
  ];

  const inputs = new Map();
  const rows = fields.map(([key, label, value, unit, help]) => {
    const input = element('input', {
      type: 'number', value: String(value), min: '1', size: 8,
    });
    inputs.set(key, input);
    return element('label', { class: 'field explained', title: help }, [
      element('span', {}, element('span', { text: unit ? `${label} (${unit})` : label })),
      input,
    ]);
  });

  // Nothing to save until something is different. A button that is always
  // ready invites a press that does nothing, and then you cannot tell whether
  // the last press took.
  const asShown = () => {
    const now = {};
    inputs.forEach((input, key) => { now[key] = Number(input.value); });
    return now;
  };
  const unchanged = () => fields.every(([key, , value]) =>
    Number(asShown()[key]) === Number(value));

  const save = element('button', {
    class: 'action', text: 'Save', disabled: 'disabled',
    onclick: async () => {
      const changes = asShown();
      save.disabled = true;
      try {
        await api.updateGateway(changes);
        // Nothing is said on success: the fields keep the saved values and
        // Save goes back to sleep, which is the confirmation.
      } catch (error) {
        await showNotice({ title: 'Could not save the gateway limits',
                           body: error.message });
      }
      redraw();
    },
  });

  inputs.forEach((input) => {
    input.addEventListener('input', () => { save.disabled = unchanged(); });
  });

  return section('Limits', [
    ...rows,
    element('div', { class: 'row' }, [element('span', {}), save]),
  ]);
}


// The queue, as the order it will actually be served in.
//
// It is served in order and requests next to each other wanting the same model
// go in together, so it is not a list of requests — it is a list of turns. Read
// down it and you have the schedule of model changes that is about to happen,
// which is the thing worth knowing before it does.
function coming(stats) {
  const runs = stats.queue_runs || [];
  const loaded = stats.loaded || [];
  const rows = [];

  // What is on the machine, one row each: what it is answering now and how
  // many are queued for it. Both, because they say different things — running
  // is whether it is busy, waiting is whether there is pressure on it.
  loaded.forEach((item) => {
    rows.push(element('div', { class: 'row tight now' }, [
      element('span', { class: 'swap', text: `→ ${item.instance_id}` }),
      element('span', { class: 'muted', text:
        `${figure(item.in_flight)} running · ${figure(item.waiting)} waiting` }),
    ]));
  });
  if (!loaded.length) {
    rows.push(element('div', { class: 'row tight now' },
                      element('span', { class: 'muted', text: 'nothing loaded' })));
  }

  // Everything queued for a model that is already there. It is waiting for a
  // place, not for a load.
  const onLoaded = loaded.reduce((total, item) => total + (Number(item.waiting) || 0), 0);
  rows.push(element('div', { class: 'row tight tally' }, [
    element('span', { class: 'muted', text: 'Waiting' }),
    element('span', { class: 'muted', text: figure(onLoaded) }),
  ]));

  // The next model that has to be loaded, and everything held up behind it.
  // Read together they are the cost of the change: this many requests are
  // stopped until that one is on.
  const next = runs[0];
  const behind = next
    ? runs.reduce((total, run) => total + (Number(run.requests) || 0), 0)
      - (Number(next.requests) || 0)
    : 0;
  rows.push(element('div', { class: 'row tight next',
                             ...(next ? { title: 'longest has waited '
                                                 + `${next.longest_wait_s} s` } : {}) }, [
    element('span', { class: 'muted', text: 'Next change' }),
    next
      ? element('span', { class: 'swap grow', text: `${next.instance_id}  ←` })
      : element('span', { class: 'muted grow', text: '—' }),
    next
      ? element('span', { class: 'muted', text: `${figure(next.requests)} waiting` })
      : null,
  ].filter(Boolean)));
  rows.push(element('div', { class: 'row tight tally' }, [
    element('span', { class: 'muted', text: 'Remaining' }),
    element('span', { class: 'muted', text: figure(behind) }),
  ]));
  return section('Queue', rows);
}


export function stopRefreshing() {
  if (timer) { window.clearInterval(timer); timer = null; }
}

export async function render(container) {
  // One timer at a time. Leaving the old one running would make the page
  // redraw twice as often for every visit to this tab.
  stopRefreshing();

  let stats;
  try {
    stats = await api.gateway();
  } catch (error) {
    container.replaceChildren(element('p', { class: 'error', text: error.message }));
    return;
  }

  // The grid is its own element rather than the shared page container: a
  // class put on that one stays behind when another tab is opened, and the
  // Models and Library pages inherited a layout meant for this one.
  container.replaceChildren(element('div', { class: 'gateway-grid' }, [
    element('div', { class: 'at-activity' }, activity(stats)),
    element('div', { class: 'at-address' }, address(stats)),
    element('div', { class: 'at-limits' }, limits(stats, () => render(container))),
    element('div', { class: 'at-queue' }, coming(stats)),
  ]));

  // window.setInterval rather than the bare global, so closing the page
  // cancels it. A timer that outlives the window keeps the whole thing alive.
  const mine = container.firstChild;
  timer = window.setInterval(() => {
    if (container.firstChild !== mine) { window.clearInterval(timer); timer = null; return; }
    render(container).catch(() => {});
  }, EVERY_MS);
}
